core/src/main/java/org/apache/spark/shuffle/sort/io/LocalDiskShuffleExecutorComponents.java

```diff
@@ -17,69 +17,64 @@
 
 package org.apache.spark.shuffle.sort.io;
 
-import java.util.Map;
 import java.util.Optional;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
 
 import org.apache.spark.SparkConf;
 import org.apache.spark.SparkEnv;
+import org.apache.spark.shuffle.IndexShuffleBlockResolver;
 import org.apache.spark.shuffle.api.ShuffleExecutorComponents;
 import org.apache.spark.shuffle.api.ShuffleMapOutputWriter;
-import org.apache.spark.shuffle.IndexShuffleBlockResolver;
 import org.apache.spark.shuffle.api.SingleSpillShuffleMapOutputWriter;
-import org.apache.spark.storage.BlockManager;
 
 public class LocalDiskShuffleExecutorComponents implements ShuffleExecutorComponents {
 
   private final SparkConf sparkConf;
-  private BlockManager blockManager;
-  private IndexShuffleBlockResolver blockResolver;
+  private final Supplier<IndexShuffleBlockResolver> blockResolver;
```

Comment:
Here and elsewhere, why rely on `Supplier` when output is memoized ?
Replace with instance instead ?

Reply:
I believe this is a supplier because at the time the executor components are initialized, the entire SparkEnv instance is not initialized yet, so we want the block resolver to be fetched from the SparkEnv lazily.

Reply:
Can you elaborate on this ? An example where `SparkEnv` was not initialized but executor shuffle component was ?
The usages I saw all had `SparkEnv` already available, but I could have missed some flow.

Reply:
The flow I see (ignoring tests) is :
`MemoizingShuffleDataIO._executor` -> `LocalDiskShuffleDataIO.initializeShuffleExecutorComponents` -> `LocalDiskShuffleExecutorComponents ctor`

 `_executor` is a `lazy val`, but its construction requires `SparkEnv` to have been set (and so initialized).
For this path, we will have block manager from `SparkEnv` available to create `IndexShuffleBlockResolver` in `LocalDiskShuffleExecutorComponents`.

Is there any other path or flow which requires this to be lazy ?

Reply:
Hm, I think that now that the laziness is handled at the level above (MemoizingShuffleDataIO) it should be fine to make these non-lazy then. I'll try that and push and check the build after. I recall there being something weird with local mode, but that might have been a state that isn't possible now with the new memoizing layer above.